tracer/test/Datadog.Trace.IntegrationTests/Sampling/DelaySamplingDecisionTests.cs

```diff
@@ -0,0 +1,91 @@
+        var sampler = new Mock<ITraceSampler>();
+        var scopeManager = new AsyncLocalScopeManager();
+        var statsd = new NoOpStatsd();
+
+        var tracer = new Tracer(settings, agentWriter.Object, sampler.Object, scopeManager, statsd);
+        TraceContext traceContext;
+
+        using (var scope1 = (Scope)tracer.StartActive("operation"))
+        {
+            traceContext = scope1.Span.Context.TraceContext;
+
+            // sampling decision not taken when first span starts
+            sampler.Verify(s => s.MakeSamplingDecision(It.IsAny<Span>()), Times.Never);
+            traceContext.SamplingPriority.Should().BeNull();
+
+            using ((Scope)tracer.StartActive("operation"))
+            {
+                // sampling decision still not taken when second span starts
+                sampler.Verify(s => s.MakeSamplingDecision(It.IsAny<Span>()), Times.Never);
+                traceContext.SamplingPriority.Should().BeNull();
+            }
+
+            // sampling decision still not taken when second span ends
+            sampler.Verify(s => s.MakeSamplingDecision(It.IsAny<Span>()), Times.Never);
+            traceContext.SamplingPriority.Should().BeNull();
+        }
+
+        // sampling decision IS taken after LAST span ends
+        sampler.Verify(s => s.MakeSamplingDecision(It.IsAny<Span>()), Times.Once);
+        traceContext.SamplingPriority.Should().NotBeNull();
+    }
+
+    [Fact]
+    public void SamplingDecisionIsMadeWhenPropagating()
+    {
+        var settings = new TracerSettings();
+        var agentWriter = new Mock<IAgentWriter>();
+        var sampler = new Mock<ITraceSampler>();
+        var scopeManager = new AsyncLocalScopeManager();
+        var statsd = new NoOpStatsd();
+
+        var tracer = new Tracer(settings, agentWriter.Object, sampler.Object, scopeManager, statsd);
+        TraceContext traceContext;
+
+        using (var scope1 = (Scope)tracer.StartActive("operation"))
+        {
+            traceContext = scope1.Span.Context.TraceContext;
+
+            // sampling decision not taken when first span starts
+            sampler.Verify(s => s.MakeSamplingDecision(It.IsAny<Span>()), Times.Never);
+            traceContext.SamplingPriority.Should().BeNull();
+
+            var headers = new NameValueCollection();
+            SpanContextPropagator.Instance.Inject(scope1.Span.Context, headers.Wrap());
+
+            // sampling decision IS taken before propagating
+            sampler.Verify(s => s.MakeSamplingDecision(It.IsAny<Span>()), Times.Once);
+            traceContext.SamplingPriority.Should().NotBeNull();
+            headers["x-datadog-sampling-priority"].Should().Be(traceContext.SamplingPriority.ToString());
+        }
+
+        // sampling decision is still there and is NOT taken again when first span ends
+        sampler.Verify(s => s.MakeSamplingDecision(It.IsAny<Span>()), Times.Once);
+        traceContext.SamplingPriority.Should().NotBeNull();
```

Comment:
nit: I think we should compare this to the previous version to make sure it's identical

Reply:
Can do, thanks.

Reply:
Done.